Add gRPC service definition for Plugin resources

justinas opened the pull request:
Contributes to https://github.com/gravitational/cloud/issues/2506, see also RFD https://github.com/gravitational/cloud/pull/2678

Adds dedicated gRPC plugin service definitions, new preset RBAC rule to manipulate plugins, as well as a couple smaller supporting changes to the Enterprise PR.
 
The Enterprise counterpart is https://github.com/gravitational/teleport.e/pull/844 . Since the suggestion to move this into a dedicated service, most of the changes were moved into Enterprise.

Commit:
Add Plugin methods to auth

Commit:
Add gRPC-layer methods for Plugin

Commit:
Add RBAC presets for Plugin

Commit:
Test GetPlugin()/NoSecrets access

Commit:
Make error assertions more correct in role test

Commit:
Deny setting credentials if user can not read them

Commit:
Add plugin exchange service

justinas commented on lib/services/presets.go:
```diff
@@ -79,6 +79,7 @@ func NewPresetEditorRole() types.Role {
 					types.NewRule(types.KindInstance, RO()),
 					types.NewRule(types.KindLoginRule, RW()),
 					types.NewRule(types.KindSAMLIdPServiceProvider, RW()),
+					types.NewRule(types.KindPlugin, append(ReadNoSecrets(), types.VerbCreate, types.VerbUpdate, types.VerbDelete)),
```
My reasoning here is that since auth exchanges the initial credentials, and the plugin (which interacts with auth server directly in this case) renews them, there is no reason to let even `editor` manipulate those.

This may or may not be overkill, let me know.

espadolini replied:
I'll let someone else chime in on RW vs RO-plus-select-edits (it seems used for `KindDevice` at least, so it wouldn't be new) and I'll just say that it feels very weird for `editor` to only have `ReadNoSecrets`.

justinas replied:
So `Plugin` stores OAuth credentials associated with us (the app) & the user. If/when this is extended to self-hosted Enterprise customers, they would be setting up their OAuth app anyway. In Cloud however, these would be credentials issued for a "Teleport Cloud" app managed by us, meaning if a user can get the credentials, they can impersonate this app with our branding, albeit only in their own Slack workspace / PagerDuty account / etc.. 🙂 That's the reason I started with conservative permissions.

Would it be viable to change this preset to RW later on as needed?

espadolini replied:
If that's a concern then the permission system is not the correct place where to put this restriction, it should be made structural.

Nothing here stops me from adding all  permissions for the plugin type (or any type, I typically just give my user permissions for the `*` verb on the `*` resource).

justinas replied:
We can probably make this `RW()`, and remove the ability to read secrets via the gRPC layer then? The plugins themselves will skip gRPC and access the service directly.

Commit:
gofmt

tigrato commented on api/client/client.go:
```diff
@@ -3185,3 +3185,75 @@ func (c *Client) DeleteAllSAMLIdPServiceProviders(ctx context.Context) error {
+func (c *Client) SetPluginCredentials(ctx context.Context, name string, creds types.PluginCredentials) error {
+	v1, ok := creds.(*types.PluginCredentialsV1)
+	if !ok {
+		return trace.BadParameter("unsupported plugin credentials type %T", creds)
```
```suggestion
		return trace.BadParameter("unsupported plugin credentials. expected *types.PluginCredentialsV1; got %T", creds)
```

tigrato commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -1446,6 +1446,32 @@ message DeleteSAMLIdPServiceProviderRequest {
+  // InitialCredentials are the initial credentials of the plugin.
+  // In the scope of processing this request, these are exchanged for
+  // "live" credentials, which are stored in the Plugin.
+  types.PluginInitialCredentialsV1 InitialCredentials = 2 [(gogoproto.jsontag) = "initial_credentials"];
```
Maybe renaming it to Bootstrap credentials makes it more clear

tigrato commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -2715,4 +2741,25 @@ service AuthService {
   rpc DeleteAllSAMLIdPServiceProviders(google.protobuf.Empty) returns (google.protobuf.Empty);
+
+  // CreatePlugin creates a new plugin instance.
+  rpc CreatePlugin(CreatePluginRequest) returns (google.protobuf.Empty);
```
Can't we move these methods into a PluginService so we don't add more stuff into AuthService?

justinas replied:
Will try to move to a dedicated service a la `loginrule`.

tigrato commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -1446,6 +1446,32 @@ message DeleteSAMLIdPServiceProviderRequest {
+message SetPluginCredentialsRequest {
+  // Name is the name of the plugin instance.
+  string Name = 1 [(gogoproto.jsontag) = "name"];
+  // Credentials are the live plugin credentials.
```
Maybe adding something more clear such as:
`are the credentials obtained after exchanging the bootstrap credentials and successive renewals`

tigrato commented on go.mod:
```diff
@@ -167,6 +167,8 @@ require (
 	github.com/mailgun/minheap v0.0.0-20170619185613-3dbe6c6bf55f // indirect
 )
 
+require github.com/gravitational/teleport-plugins v0.0.0-20230210191358-1defd181af97
```
can you move this block into any require block above?

justinas replied:
> Should we be pointing at a tag instead of a random commit?

This has the same problem as `gravitational/teleport/api`: we tag major versions, but do not change the import path as Go expects:

```console
$ go get github.com/gravitational/teleport-plugins@v12.0.2
go: github.com/gravitational/teleport-plugins@v12.0.2: invalid version: module contains a go.mod file, so module path must match major version ("github.com/gravitational/teleport-plugins/v12")
```

tigrato commented on lib/auth/auth.go:
```diff
@@ -3962,6 +3976,94 @@ func (a *Server) DeleteKubernetesCluster(ctx context.Context, name string) error
 
+// CreatePlugin creates a new plugin instance.
+func (a *Server) CreatePlugin(ctx context.Context, req *proto.CreatePluginRequest) error {
+	svc := a.pluginExchangeService
```
should this be guarded by the same lock used in `SetPluginExchangeService`?

justinas replied:
Now that I looked at it better, other Enteprise services, e.g. SAMLService, do not use the lock at all:

https://github.com/gravitational/teleport/blob/e3ce1f83c45db68fd5b8c8f633a22d627d337bb1/lib/auth/auth.go#L536-L541

So it might be superfluous here, since we only initialize this once on startup.

tigrato commented on lib/auth/plugins.go:
```diff
@@ -0,0 +1,32 @@
+package auth
+
+import (
+	"github.com/gravitational/teleport-plugins/access/common/auth/oauth"
```
Do you plan to import other things besides that interface?
If the answer is no, we probably can drop the import locally since it's just importing an interface def and we prefer local interfaces

justinas replied:
`teleport.e` will import more things from `teleport-plugins` (the concrete plugin types themselves)

tigrato commented on lib/auth/auth.go:
```diff
@@ -3962,6 +3976,94 @@ func (a *Server) DeleteKubernetesCluster(ctx context.Context, name string) error
+// GetPlugin returns a plugin instance by name.
+func (a *Server) GetPlugin(ctx context.Context, name string, withSecrets bool) (types.Plugin, error) {
+	plugin, err := a.Services.Plugins.GetPlugin(ctx, name, withSecrets)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return plugin, nil
```
```suggestion
	return plugin, trace.Wrap(err)
```

tigrato commented on lib/auth/auth.go:
```diff
@@ -3962,6 +3976,94 @@ func (a *Server) DeleteKubernetesCluster(ctx context.Context, name string) error
+// GetPlugins returns all plugin instances.
+func (a *Server) GetPlugins(ctx context.Context, withSecrets bool) ([]types.Plugin, error) {
+	plugin, err := a.Services.Plugins.GetPlugins(ctx, withSecrets)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return plugin, nil
```
```suggestion
	return plugin, trace.Wrap(err)
```

tigrato commented on lib/auth/auth_with_roles.go:
```diff
@@ -5481,6 +5481,95 @@ func (a *ServerWithRoles) DeleteAllSAMLIdPServiceProviders(ctx context.Context)
+
+// GetPlugins returns all plugin instances.
+func (a *ServerWithRoles) GetPlugins(ctx context.Context, withSecrets bool) ([]types.Plugin, error) {
+	if err := a.action(apidefaults.Namespace, types.KindPlugin, types.VerbList); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := a.action(apidefaults.Namespace, types.KindPlugin, types.VerbReadNoSecrets); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if withSecrets {
+		if err := a.action(apidefaults.Namespace, types.KindPlugin, types.VerbRead); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
```
```suggestion
	readVerb := types.VerbReadNoSecrets
	if withSecrets {
		readVerb = types.VerbRead
	}
	if err := a.action(apidefaults.Namespace, types.KindPlugin, types.VerbList, readVerb); err != nil {
		return nil, trace.Wrap(err)
	}
```

Would something like this work?

tigrato commented on lib/auth/auth_with_roles.go:
```diff
@@ -5481,6 +5481,95 @@ func (a *ServerWithRoles) DeleteAllSAMLIdPServiceProviders(ctx context.Context)
+
+// GetPlugin returns a plugin instance by name.
+func (a *ServerWithRoles) GetPlugin(ctx context.Context, name string, withSecrets bool) (types.Plugin, error) {
+	if err := a.action(apidefaults.Namespace, types.KindPlugin, types.VerbReadNoSecrets); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if withSecrets {
+		if err := a.action(apidefaults.Namespace, types.KindPlugin, types.VerbRead); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
```
```suggestion
	readVerb := types.VerbReadNoSecrets
	if withSecrets {
		readVerb = types.VerbRead
	}
	if err := a.action(apidefaults.Namespace, types.KindPlugin, readVerb); err != nil {
		return nil, trace.Wrap(err)
	}
```

Would something like this work?

tigrato commented on lib/auth/grpcserver.go:
```diff
@@ -4744,6 +4744,110 @@ func (g *GRPCServer) DeleteAllSAMLIdPServiceProviders(ctx context.Context, _ *em
+	}
+	v1, ok := plugin.(*types.PluginV1)
+	if !ok {
+		return nil, trace.Errorf("unexpected plugin type: %T", plugin)
```
Identify what do you expect to receive here

tigrato commented on lib/auth/grpcserver.go:
```diff
@@ -4744,6 +4744,110 @@ func (g *GRPCServer) DeleteAllSAMLIdPServiceProviders(ctx context.Context, _ *em
+	for _, plugin := range plugins {
+		pluginV1, ok := plugin.(*types.PluginV1)
+		if !ok {
+			return nil, trace.BadParameter("unsupported plugin type %T", plugin)
```
Same as above

zmb3 commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -3881,3 +3881,173 @@ func TestGetLicensePermissions(t *testing.T) {
+			user, err := CreateUser(srv.Auth(), "test-user", role)
+			require.NoError(t, err)
+
+			client, err := srv.NewClient(TestUser(user.GetName()))
```
Are we missing a `Close()` on client?

justinas commented on lib/auth/plugins.go:
```diff
@@ -0,0 +1,32 @@
+	// for the given plugin according to its type.
+	// If the exchanger can not be found (e.g. exchanger for this type is not configured),
+	// nil is returned.
+	GetExchanger(types.Plugin) oauth.Exchanger
```
1. We could fathomably add something like:
```go
type PluginType int
const (
    PluginTypeUnknown PluginType = 0
    PluginTypeSlackAccess = 1
)

type Plugin interface {
    // <...>
    Type() PluginType
}
```

and make GetExchanger operate on that:

```
GetExchanger(types.PluginType) oauth.Exchanger
```

2. Yes, probably a good idea to have a more descriptive error return rather than just `nil`.

justinas committed:
Apply minor suggestions from code review

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>
Co-authored-by: Tiago Silva <tiago.silva@goteleport.com>

justinas commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -2715,4 +2741,25 @@ service AuthService {
+  rpc GetPlugin(types.ResourceWithSecretsRequest) returns (types.PluginV1);
+
+  // GetPlugins returns all plugin instances.
+  rpc GetPlugins(types.ResourcesWithSecretsRequest) returns (types.PluginListV1);
```
This will be consumed by the UI. However, we expect a low amount of plugins in a cluster (~single digit). In fact, for the very first implementation, there will only be 1 type (Slack), and there will not be an option to connect multiple instances for one type. So the maximum amount can only be 1 plugin for now.

Please let me know if I should implement `ListPlugins` regardless for future-proofing.

Commit:
Move dependency into an existing block in go.mod

Commit:
Improve error messages for failed type assertions

Commit:
DRY WithSecrets access checks for Plugins

Commit:
Run new tests in parallel

Commit:
Improve error assertions in auth_with_roles_test

Commit:
TestGetPluginWithSecrets: split cases to subtests

Commit:
Clean up test servers and clients

Commit:
Add proto for plugin service

Commit:
Remove Plugin methods from auth service

Moved to a dedicated service

Commit:
Remove plugin-related auth methods

Moved to a dedicated service in Enterprise

Commit:
Remove CreatePlugin test from auth_with_roles_test

Moved to a dedicated service in Enterprise

Commit:
Pass "backend getter" to local plugins service

This pattern is used in Enterprise to set up secondary services
before auth (and backend) are created.

Commit:
Rename InitialCredentials to BootstrapCredentials

Commit:
Merge remote-tracking branch 'origin/master' into justinas/plugin-auth-methods

Commit:
Add plugins service to genproto.sh

Commit:
Reformat generated proto

justinas commented:
@tigrato @rosstimothy I moved the implementation into a dedicated service (in the Enterprise PR), and applied most of the suggestions I think. Please take a look at this and the Enterprise PR.

Commit:
Remove obsolete PluginExchangeService

The equivalent of this is now in Enterprise

tigrato commented on lib/services/local/plugins.go:
```diff
@@ -32,11 +32,15 @@ const (
 // PluginsService manages plugin instances in the backend.
 type PluginsService struct {
-	backend backend.Backend
+	getBackend func() backend.Backend
```
why do you need to use it as a function?

justinas replied:
This is an established pattern used in Enterprise/Cloud addons, where the initialization of these addons happens before the Auth (and its backend) is initialized fully.

See the Enterprise PR: https://github.com/gravitational/teleport.e/pull/844/files#diff-d4ea8e373e1f92f375929cfe3e9b2e4cb93700829fe7fef7689c552a0add6f3aR230

and an existing example of such code: https://github.com/gravitational/teleport.e/blob/82681e1ce14ad1ae321a8d31b4baf8e1b8f15e46/lib/auth/plugin.go#L129

tigrato replied:
Got it

Commit:
Add kube service to genproto.sh; regenerate

Commit:
Merge remote-tracking branch 'origin/master' into justinas/plugin-auth-methods

Commit:
Add ListPlugins to plugin backend service

Commit:
Reimplement GetPlugins on top of ListPlugins

This is a "convenience" implementation for the backend service layer.

Commit:
Replace GetPlugins with ListPlugins in gRPC schema

tigrato committed:
Fix ListKubernetesResources unit test

espadolini commented on lib/auth/auth_with_roles.go:
```diff
@@ -266,6 +267,12 @@ func (a *ServerWithRoles) LoginRuleClient() loginrulepb.LoginRuleServiceClient {
+// PluginsClient allows ServerWithRoles to implement ClientI.
+// It should not be called through ServerWithRoles and will always panic.
+func (a *ServerWithRoles) PluginsClient() pluginspb.PluginServiceClient {
+	panic("PluginsClient not implemented by ServerWithRoles")
```
This feels really bad; could we replace this one (and the other two `FooClient` functions) with a `pluginspb.PluginServiceClient` that always returns `NotImplemented`? It could be done for all of them in a single pass by implementing a noop `grpc.ClientConnInterface` that always errors out.

espadolini commented on lib/backend/backend.go:
```diff
@@ -301,6 +301,8 @@ func NextPaginationKey(r types.Resource) string {
 		return string(nextKey(internalKey(resourceWithType.GetHostID(), resourceWithType.GetName())))
 	case types.KubeServer:
 		return string(nextKey(internalKey(resourceWithType.GetHostID(), resourceWithType.GetName())))
+	case types.Plugin:
+		return string(nextKey(Key("plugins", resourceWithType.GetName())))
```
We only have a single field to query on, why not leave it unprefixed?

espadolini commented on lib/services/local/plugins.go:
```diff
@@ -74,7 +77,7 @@ func (s *PluginsService) DeletePlugin(ctx context.Context, name string) error {
 // DeleteAllPlugins implements service.Plugins
 func (s *PluginsService) DeleteAllPlugins(ctx context.Context) error {
 	startKey := backend.Key(pluginsPrefix)
-	err := s.backend.DeleteRange(ctx, startKey, backend.RangeEnd(startKey))
+	err := s.backend().DeleteRange(ctx, startKey, backend.RangeEnd(startKey))
```
Just in case we end up adding a resource that's just a singleton called `/plugint` (😅), this should use `startKey := backend.Key(pluginsPrefix, "")`

espadolini commented on lib/services/local/plugins.go:
```diff
@@ -104,25 +107,56 @@ func (s *PluginsService) GetPlugin(ctx context.Context, name string, withSecrets
 func (s *PluginsService) GetPlugins(ctx context.Context, withSecrets bool) ([]types.Plugin, error) {
-	startKey := backend.Key(pluginsPrefix)
-	result, err := s.backend.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
+	const pageSize = 1000
```
`apidefaults.DefaultChunkSize`?

espadolini commented on lib/services/local/plugins.go:
```diff
@@ -104,25 +107,56 @@ func (s *PluginsService) GetPlugin(ctx context.Context, name string, withSecrets
+// ListPlugins returns a paginated list of plugin instances.
+// StartKey is a resource name, which is the suffix of its key.
+func (s *PluginsService) ListPlugins(ctx context.Context, limit int, startKey string, withSecrets bool) ([]types.Plugin, string, error) {
+	startKeyBytes := []byte(startKey)
```
Relying on the user to specify the correct prefix in the `startKey` can cause confusion at best, security issues at worst. The `startKey` should just be the resource name and here we can just `backend.Key(pluginsPrefix, startKey)`.

espadolini commented on lib/services/local/plugins.go:
```diff
@@ -104,25 +107,56 @@ func (s *PluginsService) GetPlugin(ctx context.Context, name string, withSecrets
-	return plugins, nil
+	var nextKey string
+	if len(result.Items) == limit {
+		nextKey = backend.NextPaginationKey(plugins[len(plugins)-1])
```
`startKey` is inclusive, so this will cause the item at the very end of the page to be seen twice.

justinas replied:
`NextPaginationKey` uses `nextKey` to increment by 1, so this should behave correctly? `TestListPlugins` should also verify this.

espadolini replied:
You're right; in a lot of places we fetch `limit + 1` items when paginating just so the client doesn't have to do one extra API call if the page limit hits the last item exactly, but maybe we don't care for this specific case.

espadolini replied:
quoting @rosstimothy: we should probably do pagination like we do in the other places, and that would be by fetching `limit + 1` items and using the name of the extra one as the `nextKey`; if we want to change how pagination should behave for the one-past-the-last page, we should do that consistently

Commit:
Simplify plugin pagination key to just the name

Commit:
Use existing constant for page size

justinas commented on go.mod:
```diff
@@ -74,6 +74,7 @@ require (
 	github.com/gravitational/oxy v0.0.0-20221029012416-9fbf4c444680
 	github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
 	github.com/gravitational/roundtrip v1.0.2
+	github.com/gravitational/teleport-plugins v0.0.0-20230210191358-1defd181af97
```
`teleport.e` will end up using the plugins code quite a bit (both interfaces, and concrete types). The best solution is probably the merge of `teleport` and `teleport-plugins` - this idea has been floated a few times, but no concrete plans exist yet AFAIK.

Commit:
Make dummy clients return errors instead of panic

Commit:
Remove obsolete field

Commit:
Ensure go.mod is valid for corresponding e changes

tigrato committed:
Fix passing mutex ref

espadolini commented on lib/utils/grpc.go:
```diff
@@ -128,3 +129,17 @@ func ChainStreamServerInterceptors(first grpc.StreamServerInterceptor, rest ...g
+type GRPCDummyClientConnection struct{}
+
+// Invoke implements grpc.ClientConnInterface
+func (*GRPCDummyClientConnection) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	return trace.NotImplemented("Invoke() called on GRPCDummyClientConnection")
+}
+
+// NewStream implements grpc.ClientConnInterface
+func (*GRPCDummyClientConnection) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return nil, trace.NotImplemented("NewStream() called on GRPCDummyClientConnection")
+}
```
Maybe `GRPCDummyClientConnection` could hold an error that we `trace.WrapWithMessage`?

justinas commented on lib/services/plugins.go:
```diff
@@ -25,6 +25,12 @@ import (
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/utils"
+
+	// Dummy import so that `go mod tidy` does not clean up this module
+	// that is referenced in teleport.e, but not teleport.
+	// TODO(justinas): remove this import when bumping the e-ref
+	// after the merge of teleport.e#844.
+	_ "github.com/gravitational/teleport-plugins/access/common"
```
For sure, wasn't aware of that one 🙂

Commit:
Move teleport-plugins import to e_imports

Commit:
Revert oauth change in go.mod

justinas commented on go.mod:
```diff
@@ -127,10 +128,10 @@ require (
 	go.opentelemetry.io/proto/otlp v0.19.0
 	golang.org/x/crypto v0.6.0 // replaced
-	golang.org/x/exp v0.0.0-20221126150942-6ab00d035af9
+	golang.org/x/exp v0.0.0-20230116083435-1de6713980de
 	golang.org/x/mod v0.8.0
 	golang.org/x/net v0.7.0
-	golang.org/x/oauth2 v0.5.0
+	golang.org/x/oauth2 v0.4.0
```
`oauth` seems to revert successfully, `exp` however is auto-upgraded by `go mod tidy`, seems like `teleport-plugins` rely on this newer version.

Commit:
Use limit+1 to look-ahead when paginating plugins

Commit:
Test plugin pagination with pageSize > numPlugins

Commit:
Add descriptive messages to gRPC dummy clients

espadolini commented on lib/services/local/plugins.go:
```diff
@@ -104,25 +108,58 @@ func (s *PluginsService) GetPlugin(ctx context.Context, name string, withSecrets
+	var page []types.Plugin
+	var startKey string
+	var err error
+	for {
```
Add some maximum iteration limit here, just in case something terrible happens?

espadolini reviewed (approved):
LGTM other than deciding what the `editor` permissions should be for `KindPlugin`.

Commit:
Merge remote-tracking branch 'origin/master' into justinas/plugin-auth-methods

Commit:
Merge remote-tracking branch 'origin/master' into justinas/plugin-auth-methods

Commit:
Plugin: add RW for editor; remove secrets from gRPC

Commit:
Make message more descriptive for dummy gRPC conn

justinas commented on lib/utils/grpc.go:
```diff
@@ -128,3 +129,25 @@ func ChainStreamServerInterceptors(first grpc.StreamServerInterceptor, rest ...g
+// NewStream implements grpc.ClientConnInterface
+func (g *grpcDummyClientConnection) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return nil, trace.NotImplemented(g.message)
+}
```
Added in d423c5ad:

```
/teleport.loginrule.v1.LoginRuleService/ListLoginRules: LoginRuleClient() should not be called on ServerWithRoles
```
